torch/csrc/autograd/input_buffer.cpp

```diff
@@ -129,96 +149,137 @@ static void accumulate(
-      if (opt_sync_stream && (opt_accumulate_stream != opt_sync_stream)) {
-        // (3b), (4b)
-        c10::OptionalDeviceGuard device_guard{opt_sync_stream->device()};
-        auto event = c10::Event{device_type};
-        event.record(*opt_sync_stream);
-        opt_accumulate_stream->wait(event);
-        const auto guard = c10::impl::VirtualGuardImpl(device_type);
-        record_stream_any_impl(var, *opt_accumulate_stream);
+        // Case 4/5
+        if (opt_producer_stream->device() == *device) {
+          opt_accum_streams[pos] = opt_producer_stream;
+        } else {
+          opt_accum_streams[pos] =
+              at::accelerator::getCurrentStream(device->index());
+        }
       }
+      mb_record_stream_any_impl(
+          var, *opt_accum_streams[pos], /*prev_stream=*/*opt_producer_stream);
     }
-  }
-
-  auto& old_var = buffer[pos];
-  if (!old_var.defined()) {
+    TORCH_INTERNAL_ASSERT(!buffer[pos].defined())
     buffer[pos] = std::move(var);
-  } else {
-    if (opt_accumulate_stream) {
-      c10::OptionalStreamGuard stream_guard{opt_accumulate_stream};
+    return;
+  }
+  // [ Nth producer ]
+  // At this point, the accumulation stream has been determined. For both 2/3
+  // and 4/5, everytime subsequent producers call InputBuffer.add:
+  // - The accumulation stream (determined at n=1) waits for the new
+  //   producer stream
+  // - Call var.record_stream(accumution_stream)
+  // - Accumulate var into buffer
+  //
+  // Also, depending on whether you are in case 2/3 or 4/5, you must do
+  // something extra:
+  // - If you are in the 2/3 regime, you skipped synchronizing the first
+  //   producer with the consumer during the first iteration. We make this
+  //   up in the second iteration.
+  // - If you are in the 4/5 regime, all you've done so far is synchronize
+  //   between producers. You still need to have the consumer wait for the
+  //   producer.
+  auto accum_stream = opt_accum_streams[pos];
+  if (is_accelerator) {
+    // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+    if (opt_consumer_stream->device() == *device) {
+      // Case 2/3
+      if (current_index == 1) {
+        // 2/3 extra logic
+        mb_wait_stream(
+            accum_stream, opt_first_producer_streams[pos], device_type);
+      }
+      mb_wait_stream(accum_stream, opt_producer_stream, device_type);
+      mb_record_stream_any_impl(
+          var, *accum_stream, /*prev_stream=*/*opt_producer_stream);
+      c10::OptionalStreamGuard stream_guard{accum_stream};
       accumulate(buffer, pos, std::move(var));
     } else {
-      // (1) non-CUDA/privateuse1 variable
-      //     Accumulation happens on variable's device
-      c10::OptionalDeviceGuard device_guard{device};
-      accumulate(buffer, pos, std::move(var));
+      // Case 4/5
+      mb_wait_stream(accum_stream, opt_producer_stream, device_type);
+      mb_record_stream_any_impl(
+          var, *accum_stream, /*prev_stream=*/*opt_producer_stream);
```

Comment:
this can be deduped, we always want to have the accum wait for the new producer in both the 2/3 and 4/5 cases.